[Debug Info] Emit -gdwarf-types debug info for Builtin.FixedArray<>

adrian-prantl opened the pull request:
This commit also changes how specialized types are being emitted. Previously we
would not emitthe detailed member information in the specialized type itself,
and instead rely on the fact that it was present in the unspecialized type,
however, this wold prevent us from emitting any bound generic members, so we're
now emitting the members in both variants of the type.

adrian-prantl committed:
Remove redundant parameter (NFC)

adrian-prantl commented:
@swift-ci test source compatibility

adrian-prantl commented:
@swift-ci test

augusto2112 commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1757,6 +1766,9 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     // to emit the (target!) size of the underlying basic type.
+    if (MangledName == "$exq_BVD" ||
+        MangledName == "$es4SlabVy$0_4main8MySpriteVGD")
+      llvm::errs();
```
You forgot this here

augusto2112 commented on test/DebugInfo/value-generics-embedded.swift:
```diff
@@ -1,16 +1,24 @@
-// RUN: %target-swift-frontend %s -target %target-cpu-apple-macos14 -emit-ir -g -enable-experimental-feature ValueGenerics -enable-experimental-feature Embedded -wmo -disable-availability-checking -o - | %FileCheck %s
+// RUN: %target-swift-frontend %s -target %target-cpu-apple-macos14 -emit-ir -gdwarf-types -enable-experimental-feature ValueGenerics -enable-experimental-feature Embedded -wmo -disable-availability-checking -o - | %FileCheck %s
```
embedded + `-g` should emit it as `-gdwarf-types`, I think this should stay as `-g` to test this behavior.

adrian-prantl replied:
did you implement this logic in the driver only or also in the frontend? I think it may be necessary because this is the latter.

adrian-prantl replied:
I'm taking this back. It's not.

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/9970
@swift-ci test macOS

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/9970
@swift-ci test linux

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test macOS

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test source compatibility

adrian-prantl commented:
Renamed `Slab`.

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
Linux test failure is unrelated.

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test windows

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test  linux

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test linux

adrian-prantl committed:
[Debug Info] Emit -gdwarf-types debug info for Builtin.FixedArray<>

This commit also changes how specialized types are being emitted. Previously we
would not emitthe detailed member information in the specialized type itself,
and instead rely on the fact that it was present in the unspecialized type,
however, this wold prevent us from emitting any bound generic members, so we're
now emitting the members in both variants of the type.

This uncovered a with type aliases that this commit also addresses: Because we
don't canonicalize types prior to caching in order to preserve type sugar,
alternative representations of recursive types (with one or more levels of
recursion unfolded) could create potential infinite chains of types. This is
addressed by checking whether a sugared type has an already emitted canonical
representation first, and if yes, creating a typedef node pointing directly to
it.

The donwside of doing this is that it can lead to the disappearnce of type
aliases definitions when they are used as parameters in bound generic
types. However, we still preserve that a type was `MyClass<MyAlias>`. We just
might have a typedef pointing director from `MyClass<MyAlias>` ->
`MyClass<CanonicalType>`.

rdar://144315592

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
@swift-ci test linux

adrian-prantl commented:
@swift-ci test linux

adrian-prantl committed:
[Debug Info] Avoid emitting a single fragment that covers the entire variable.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test source compatibility
